crates/shielder-relayer/src/quote.rs

```diff
@@ -1,35 +1,120 @@
+async fn _quote_fees(
+    app_state: AppState,
+    query: QuoteFeeQuery,
+) -> Result<QuoteFeeResponse, String> {
+    // Gas-related calculations.
+    let gas_price = U256::from(get_gas_price(&app_state).await?);
+    let required_gas = U256::from(app_state.relay_gas);
+    let gas_cost_native = required_gas * gas_price;
+
+    // Actual cost of performing the relay.
+    let relayer_cost_native = gas_cost_native + query.pocket_money;
+
+    // Relay commission.
+    let commission_native = relayer_cost_native * U256::from(15) / U256::from(100);
+
+    // Total cost for the user.
+    let total_cost_native = relayer_cost_native + commission_native;
+
+    // Token conversion.
+    let prices = match query.fee_token {
+        TokenKind::Native => {
+            let price = get_native_token_price(&app_state)?;
+            Prices {
+                fee_token_price: price,
+                native_token_price: price,
+                ratio: Decimal::ONE,
+            }
         }
-    }
+        TokenKind::ERC20(address) => get_token_price(&app_state, address)?,
+    };
+
+    Ok(QuoteFeeResponse {
+        total_fee: app_state.total_fee,
+        base_fee: gas_cost_native,
+        relay_fee: app_state.total_fee.saturating_sub(gas_cost_native),
+
+        total_cost_native,
+        total_cost_fee_token: scale_u256(total_cost_native, prices.ratio)?,
+
+        gas_price,
+        gas_cost_native,
+        gas_cost_fee_token: scale_u256(gas_cost_native, prices.ratio)?,
+
+        commission_native,
+        commission_fee_token: scale_u256(commission_native, prices.ratio)?,
+
+        native_token_price: prices.native_token_price,
+        fee_token_price: prices.fee_token_price,
+        token_price_ratio: prices.ratio,
+    })
+}
+
+async fn get_gas_price(app_state: &AppState) -> Result<u128, String> {
+    let provider = create_simple_provider(&app_state.node_rpc_url)
+        .await
+        .map_err(|err| format!("Failed to create provider: {err}"))?;
+
+    provider
+        .get_gas_price()
+        .await
+        .map_err(|err| format!("Failed to get gas price: {err}"))
+}
+
+struct Prices {
+    fee_token_price: Decimal,
+    native_token_price: Decimal,
+    ratio: Decimal,
+}
+
+fn get_native_token_price(app_state: &AppState) -> Result<Decimal, String> {
+    Ok(app_state
+        .prices
+        .price(app_state.native_token)
+        .ok_or("Native token price not available")?)
+}
+
+fn get_token_price(app_state: &AppState, address: Address) -> Result<Prices, String> {
+    let native_token_price = get_native_token_price(app_state)?;
+
+    let fee_token = app_state
+        .token_config
+        .iter()
+        .find(|config| config.token_address().is_some_and(|a| a == address))
+        .ok_or("Fee token not found")?;
+    let fee_token_price = app_state
+        .prices
+        .price(fee_token.coin)
+        .ok_or("Fee token price not available")?;
+
+    let ratio = native_token_price / fee_token_price;
```

Comment:
I'm wondering, the prices take decimals into account or not?

So, is price of ETH something like 2000*10^{-18} or rather 2000? 

Depending on the answer, this is either correct but weird, or incorrect :P

Reply:
discussed off-GH: I'll prepare a fix to the price feed